### PR TITLE
Check ${TF_DATA_DIR} when looking up Terraform state files

### LINKED ISSUE #838: Terragrunt always reinitializes state if TF_DATA_DIR is specified
Terragrunt does not know that `terraform.tfstate` can be stored in a path specified in `TF_DATA_DIR` environment variable, and that causes it to call `terraform init` every time.

How to reproduce:

```
TF_DATA_DIR=somewhere terragrunt plan
TF_DATA_DIR=somewhere terragrunt plan
```

### PR DESCRIPTION
**Pull request by misha-tectonic**
This prevents incessant reinitialization on every invocation of Terragrunt if
TF_DATA_DIR is set

Closes #838

### CODE REVIEW

File: cli/cli_app.go [hunk 1]
```diff
@@ -589,6 +590,9 @@ func needsInit(terragruntOptions *options.TerragruntOptions, terragruntConfig *c
 // Returns true if we need to run `terraform init` to download providers
 func providersNeedInit(terragruntOptions *options.TerragruntOptions) bool {
 	providersPath := util.JoinPath(terragruntOptions.WorkingDir, ".terraform/plugins")
+	if tfDataDir, ok := os.LookupEnv("TF_DATA_DIR"); ok {
+		providersPath = util.JoinPath(tfDataDir, "plugins")
```
Comment (brikis98) [inline cli/cli_app.go]: Should this be `plugins` or `.terraform/plugins`?

File: test/integration_test.go [hunk 2]
```diff
@@ -1875,6 +1876,45 @@ func TestDependencyOutputCachePathBug(t *testing.T) {
 	require.NoError(t, err)
 }
 
+func TestDataDir(t *testing.T) {
```
Comment (brikis98) [inline test/integration_test.go]: Thanks for adding a test! 👍

File: cli/cli_app.go [hunk 3]
```diff
@@ -661,6 +665,9 @@ func runMultiModuleCommand(command string, terragruntOptions *options.Terragrunt
 // complicated and not something we handle at the moment.
 func modulesNeedInit(terragruntOptions *options.TerragruntOptions) (bool, error) {
 	modulesPath := util.JoinPath(terragruntOptions.WorkingDir, ".terraform/modules")
+	if tfDataDir, ok := os.LookupEnv("TF_DATA_DIR"); ok {
+		modulesPath = util.JoinPath(tfDataDir, "modules")
```
Comment (brikis98) [inline cli/cli_app.go]: Seems like a lot of copy/pasted logic. Perhaps extract into a method called something like `withTfDataDir(pathParts ...string)`? If `TF_DATA_DIR` isn't set, it returns `util.JoinPath` on `pathParts`. If it is set, it joins the data dir to `pathParts`, dropping `.terraform` if it's the first arg to `pathParts`.

File: cli/cli_app.go (hunk 1 above)
Comment (misha-tectonic) [reply cli/cli_app.go]: `plugins`.

File: cli/cli_app.go (hunk 3 above)
Comment (misha-tectonic) [reply cli/cli_app.go]: Turns out every time `.terraform` is used, it comes from `TerragruntOptions`, so I have moved the logic there.

**Comment by misha-tectonic** [comment]
Done

**Comment by brikis98** [comment]
Thanks! Kicking off tests now.

**Comment by misha-tectonic** [comment]
Failing test seems to be unrelated to my change.

**Comment by brikis98** [comment]
OK, re-ran tests and they passed, so looks like it was some transient issue with S3. Merging now, thanks!

**Comment by brikis98** [comment]
https://github.com/gruntwork-io/terragrunt/releases/tag/v0.20.1

**Comment by misha-tectonic** [comment]
Thanks. Are binaries generated asynchronously, should I wait for them to appear on the releases page?

**Comment by brikis98** [comment]
It takes the build ~10 min to run and upload the binaries. They should be there shortly.

**Pull request merged**